R/class.R

```diff
@@ -56,20 +72,25 @@ PyClass <- function(classname, defs = list(), inherit = NULL) {
   defs <- lapply(defs, function(x) {
     if (inherits(x, "function")) {
       f <- inject_super(x)
-      x <- function(...) {
-        args <- list(...)
-        # enable convertion scope for `self`
-        # the first argument is always `self`.and we don't want to convert it.
-        assign("convert", TRUE, envir = as.environment(args[[1]])) 
-        do.call(f, append(args[1], lapply(args[-1], py_to_r)))
-      }
+      x <- enable_convert_scope(f)
+      attr(x, "__env__") <- environment(f)
```

Comment:
Do we really need to store a reference to the environment? It seems like we don't actually use it (beyond checking its existence).

Reply:
we actually use it  to assign the `__class__` variable that is required to not depend on `self` to find the base class.  The python interpreter seems to do it automatically, but only when you create the function inside the class definition. 

See https://docs.python.org/3/library/functions.html#super

> The zero argument form only works inside a class definition, as the compiler fills in the necessary details to correctly retrieve the class being defined, as well as accessing the current instance for ordinary methods.

Reply:
There may be an easier way to do it, but I couldn't come up with. The idea is that `super` must be able to find what is the `type` without referring to `self` otherwise it will report the wrong type when there's an additional level of inheritance.

Reply:
Ah, got it -- I think this is clear now. Thanks!